### PR TITLE
Kernel+Terminal: Modified state and close warnings for background/foreground processes in Terminal + /proc/{pid}/children to make this easier 

### PR DESCRIPTION
**Pull request by MacDue**
This is my attempt at solving #13751, and implementing a terminal close warning (similar to the one in gnome terminal, & others).

Demo:

https://user-images.githubusercontent.com/11597044/166516311-27c2ca7c-7d7a-4e94-a284-bb705df08127.mp4

When there are process(es) running in the terminal the "close modified" state will be set, and on attempting to close the terminal you'll get a warning if there's a foreground process or background processes running. (The warning indicates that it's due to background processes if no foreground process is running).

This now also adds `/proc/{pid}/children` to make it easier to lookup the child processes of a process.

### CODE REVIEW

**Comment by djwisdom** [comment]
Well what do you know? Is it SerenityOS 2.0 already? Nice attempt, looks more like resolution to me w00t! w00t! :^)

**Comment by Diogenes31** [comment]
Should we only warn if there is a foreground process running? I like to background my editor when I want to use a terminal, and it would be nice to see a warning to let me know the terminal I'm closing is still busy.

Love the PR though, and the feature makes tons of sense.

**Comment by MacDue** [comment]
> Should we only warn if there is a foreground process running? I like to background my editor when I want to use a terminal, and it would be nice to see a warning to let me know the terminal I'm closing is still busy.

I have now implemented this (along with serval varients of the close warning). Though I'm communicating the job count from the shell via `SIGUSR1`/`SIGUSR2` -- which is probably a bit of a hacky way.

**Comment by alimpfard** [comment]
You don't need to use such hacks to figure out that the shell isn't in the foreground:
```diff
diff --git a/Userland/Applications/Terminal/main.cpp b/Userland/Applications/Terminal/main.cpp
index f080476489..d91b929e20 100644
--- a/Userland/Applications/Terminal/main.cpp
+++ b/Userland/Applications/Terminal/main.cpp
@@ -286,9 +286,9 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     terminal->on_command_exit = [&] {
         app->quit(0);
     };
-    terminal->on_title_change = [&](auto title) {
-        window->set_title(title);
-    };
+    //terminal->on_title_change = [&](auto title) {
+        // window->set_title(title);
+    //};
     terminal->on_terminal_size_change = [&](auto& size) {
         window->resize(size);
     };
@@ -373,8 +373,31 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     TRY(Core::System::unveil("/etc/FileIconProvider.ini", "r"));
     TRY(Core::System::unveil("/tmp/portal/launch", "rw"));
     TRY(Core::System::unveil("/tmp/portal/config", "rw"));
+    TRY(Core::System::unveil("/proc", "r"));
     TRY(Core::System::unveil(nullptr, nullptr));
 
+    auto timer = Core::Timer::create_repeating(100, [&, previous_pid = 0]() mutable {
+        auto pid = tcgetpgrp(ptm_fd);
+        if (pid == -1) {
+            dbgln("Terminal: Failed to get terminal process group: {}", strerror(errno));
+            return;
+        }
+        if (pid == previous_pid)
+            return;
+
+        if (pid != shell_pid) {
+            auto link = Core::File::read_link(String::formatted("/proc/{}/exe", pid));
+            if (link.is_error())
+                window->set_title(String::formatted("Terminal [{}]", pid));
+            else
+                window->set_title(String::formatted("Terminal [{}] {}", pid, link.value()));
+        } else {
+            window->set_title("Terminal");
+        }
+        previous_pid = pid;
+    });
+    timer->start();
+
     window->show();
     int result = app->exec();
     dbgln("Exiting terminal, updating utmp");
```

the concept of the terminal caring what application is running in it is kinda annoying to me personally though, so I'm not exactly in love with this idea...

**Comment by Diogenes31** [comment]
> Though I'm communicating the job count from the shell via SIGUSR1/SIGUSR2 -- which is probably a bit of a hacky way. It seems to work pretty well still.

I suspect you could do something more "proper" by looking at the subprocesses of the shell in the terminal, be it through the `/proc` filesystem or some other means. Just being an ideaguy here I guess, but it's something to consider.

**Comment by Diogenes31** [comment]
> the concept of the terminal caring what application is running in it is kinda annoying to me personally though, so I'm not exactly in love with this idea...

This sounds like the sort of feature ripe for configuration, perhaps a toggle option in the settings is in order?

**Comment by MacDue** [comment]
> This sounds like the sort of feature ripe for configuration, perhaps a toggle option in the settings is in order?
I have some stashed changes that adds this as a toggle to the settings

**Comment by alimpfard** [comment]
imo, if we're gonna do this, we should make it work with any shell, not just Shell, so options:
- check `/proc/all` and look for all processes with ppid=$shell_pid
- just look for the foreground process and warn if it's not the shell
- invent some kernel mechanism and ask (the hypothetical) `/proc/$shell_pid/children` about it

**Comment by MacDue** [comment]
@alimpfard This currently this PR works with any shell for warning about the forground process. The signals I added were as an easy way to add some support for background processes too.

**Comment by Diogenes31** [comment]
Agreed, I think the best way would be through `/proc`, (I'm not super familiar with it or at a computer with serenity rn) is there a way to easily discover what subprocesses belong to a parent process? because in my mind it's as simple as checking if the shell (whatever one, doesn't have to be serenity shell) has any child processes, if so, it's "unsaved changes", otherwise it's fine to kill

**Comment by alimpfard** [comment]
@Diogenes31 sure, run the equivalent of this to find them all :P
image

**Comment by MacDue** [comment]
I'm not sure the "modified" state would work if using `/proc`. Since I don't think it sounds like a great idea to be parsing a JSON blob every 100ms or something like that (no sure).

I think you could maybe keep the signals added here in `Shell` setting the "modified" state on the terminal.
Then parse `/proc` for warning about closing the terminal.

That way you'd get the foreground and background process warnings with any shell, and also if using `Shell` you'd also get the "modified" state.

**Comment by alimpfard** [comment]
Hence the /proc/pid/children suggestion to generate and parse a significantly smaller json.
Even so, I doubt parsing that file is particularly expensive, and you can do it at a much slower rate as well

**Comment by alimpfard** [comment]
Hacking the shell this way is imo pretty much a no-go, unless we make a better communication protocol

**Comment by MacDue** [comment]
Well I'll have a look into adding /proc/pid/children, as that seems like the nicest way. Though I'd imagine /proc/pid/children to be a directory of symlinks to child processes. Which would allow simply counting them without any need for JSON.

**Comment by Diogenes31** [comment]
That looks like an optimal solution to me, though I haven't rigorously inspected the code. Also adds a very convenient feature to `/proc`

File: Userland/Applications/Terminal/main.cpp [hunk 1]
```diff
@@ -334,9 +335,44 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
+    };
+
+    auto check_terminal_quit = [&]() -> int {
+        bool has_foreground_process = tty_has_foreground_process();
+        int child_process_count = 0;
+        auto get_close_message = [&]() -> String {
+            if (has_foreground_process) {
+                return "There is still a process running in this terminal. Closing the terminal will kill it.";
+            } else if (child_process_count > 1) {
+                return String::formatted("There are background {} processes running in this terminal. Closing the terminal will kill them.", child_process_count);
```
Comment (alimpfard) [inline Userland/Applications/Terminal/main.cpp:361]:
```suggestion
                return String::formatted("There are {} background processes running in this terminal. Closing the terminal will kill them.", child_process_count);
```

File: Userland/Applications/Terminal/main.cpp [hunk 2]
```diff
@@ -334,9 +335,44 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
+            } else {
+                VERIFY(child_process_count == 1);
+                return "There is a background process running in this terminal. Closing the terminal kill it.";
+            }
+        };
+        if (has_foreground_process || (child_process_count = shell_child_process_count()) > 0)
+            return GUI::MessageBox::show(window, get_close_message(), "Close this terminal?", GUI::MessageBox::Type::Warning, GUI::MessageBox::InputType::OKCancel);
```
Comment (alimpfard) [inline Userland/Applications/Terminal/main.cpp:368]: "Close this terminal?" is a yes/no kinda question.

File: Userland/Applications/Terminal/main.cpp [hunk 3]
```diff
@@ -334,9 +335,44 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
+        return GUI::MessageBox::ExecOK;
```
Comment (alimpfard) [inline Userland/Applications/Terminal/main.cpp:369]: This factoring is a bit weird, how about something like
```c++
auto has_foreground_process = tty_has_foreground_process();
Optional<String> close_message;

if (has_foreground_process) {
    close_message = "...";
} else {
    auto child_process_count = shell_child_process_count();
    if (child_process_count > 1)
        close_message = "...";
    else if (child_process_count == 1)
        close_message = "...";
    else
        VERIFY_NOT_REACHED();
}

if (close_message.has_value())
    return GUI::MessageBox::show(..., *close_message, ...);
return GUI::MessageBox::ExecOK;
```

**Review by alimpfard — COMMENTED** [review]
This iteration is much nicer!
Some nitpicks and suggestions:

File: Userland/Applications/Terminal/main.cpp (hunk 3 above)
Comment (alimpfard) [reply Userland/Applications/Terminal/main.cpp:369]: actually, how does this now blow up when closing a normal terminal?
the assertion should blow up if there are no background processes

Comment (MacDue) [reply Userland/Applications/Terminal/main.cpp:369]: It does not blow up as `has_foreground_process` is `false` and `shell_child_process_count()`  is `< 0`. Though I agree the factoring is a bit weird.

File: Userland/Applications/Terminal/main.cpp (hunk 2 above)
Comment (MacDue) [reply Userland/Applications/Terminal/main.cpp:368]: I'm not sure, I feel like the ok/cancel reads better with the close message (not the title).

\> Closing this terminal will kill...
\> Ok

File: Userland/Applications/Terminal/main.cpp (hunk 3 above)
Comment (alimpfard) [reply Userland/Applications/Terminal/main.cpp:369]: well the code path as it stands is quite the spaghetti, please refactor it somehow to make it less so :)

File: Userland/Applications/Terminal/main.cpp (hunk 2 above)
Comment (alimpfard) [reply Userland/Applications/Terminal/main.cpp:368]: hm...not sure, let's keep it like this for now :shrug:

File: Userland/Applications/Terminal/main.cpp [hunk 4]
```diff
@@ -334,9 +335,42 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
+    auto check_terminal_quit = [&]() -> int {
+        Optional<String> close_message;
+        if (tty_has_foreground_process()) {
+            close_message = "There is still a process running in this terminal. Closing the terminal will kill it.";
+        } else {
+            auto child_process_count = shell_child_process_count();
+            if (child_process_count > 1)
+                close_message = String::formatted("There are {} background processes running in this terminal. Closing the terminal will kill them.", child_process_count);
+            else if (child_process_count == 1)
+                close_message = "There is a background process running in this terminal. Closing the terminal kill it.";
```
Comment (alimpfard) [inline Userland/Applications/Terminal/main.cpp:371]:
```suggestion
                close_message = "There is a background process running in this terminal. Closing the terminal will kill it.";
```

File: Userland/Applications/Terminal/main.cpp [hunk 5]
```diff
@@ -417,7 +425,7 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     TRY(Core::System::unveil(nullptr, nullptr));
 
     auto modified_state_check_timer = Core::Timer::create_repeating(500, [&] {
```
Comment (alimpfard) [inline Userland/Applications/Terminal/main.cpp:427]: if `should_confirm_close` is false, we'll end up with a useless timer repeating forever doing nothing, pls avoid that.

**Review by alimpfard — COMMENTED** [review]
moar nitpicks

**Pull request merged**